feat: Add verify signatures with github actions and subject prefix blog (#116)
kubewarden/kubewarden.io: content/blog/2022/07/verify-signatures-with-gha-and-prefix.md

---
title: Verify image signatures with GithubActions and KeylessPrefix 
authors:
- Raul Cabello Martin 

> date: 2022-07-14

Reviewer: Don't forget to change the date to the final one

---

With the latest Kubewarden release v1.1.0 and the latest release of
the [verify-image-signatures](https://github.com/kubewarden/verify-image-signatures) policy, it's now possible to use
`GithubActions` or `KeylessPrefix` for verifying images. Read our
previous [blog post](https://www.kubewarden.io/blog/2022/05/verifying-image-signatures/) if you want to learn more about
how to verify container images with Sigstore using Kubewarden.

## Let's see it in action!

We want to verify the image `ghcr.io/raulcabello/app-example` which was signed inside a GitHub action. Let's run the
following command to see the issuer and subject:

```

> COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/raulcabello/app-example:v0.4.0

Reviewer: If we want to keep this cli (see below) we should change it to be:

```suggestion
COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/raulcabello/app-example:v0.4.0 | jq
```

Also, don't forget to add the `console` annotation to the triple backtick

```

The output is:

```json 
{
  ....
  "Issuer": "https://token.actions.githubusercontent.com",
  "Subject": "https://github.com/raulcabello/app-example/.github/workflows/ci.yml@refs/tags/v0.4.0"
}

> ```

Reviewer: I don't feel it makes sense to talk about this output. Let's be honest, cosign is not telling you the relevant bits that are used by us when doing a GitHub verification. I'm talking about the x509 extensions that point to the GitHub owner and repository.

I would personally remove all this snippet and just say something like

> We want to verify the image `ghcr.io/raulcabello/app-example` which was built and signed inside a GitHub action
> using [this GitHub Action](https://github.com/raulcabello/app-example/blob/master/.github/workflows/ci.yml).
> 
> Out of the box, GitHub Actions have a specially crafted environment that makes Sigstore keyless signing work in
> a non-interactive way.
> The signatures produced in this way contain unique secure information that allow us to identify the GitHub owner
> (be it an individual or an organization) and the GitHub repository inside of which the GitHub Action has been
> executed.
>
> Starting from today, Kubewarden provides a convenient way to check signatures produced by GitHub actions.

And then you show the GitHub action example.

Author: done in https://github.com/kubewarden/kubewarden.io/pull/116/commits/80a27aea1eb710fffac2671ce17859f351b328c8

For this example, a Kubernetes cluster with Kubewarden already installed is required. The installation process is
described in the [quick start guide](https://docs.kubewarden.io/quick-start).

Let's verify this image with the new signature types: GitHubAction and KeylessPrefix

### GitHubAction

```yaml 
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
name: verify-image-signatures-policy
spec:
module: registry://ghcr.io/kubewarden/policies/verify-image-signatures:v0.1.4
rules:
- apiGroups: [""]
  apiVersions: ["v1"]
  resources: ["pods"]
  operations:
    - CREATE
    - UPDATE
      mutating: true
      settings:
      signatures:
        - image: "ghcr.io/raulcabello/app-example:*" # match all tags 
          github_actions:
            - owner: "raulcabello"
              repo: "app-example" #optional. if omited all apps from the owner will ve valid.
```

This policy verifies all containers with an image that is `ghcr.io/raulcabello/app-example`. It will accept
containers that were signed inside a GitHub action for the owner `raulcabello` and the app `app-example`.

> ### KeylessPrefix

Reviewer: I would not show how to do keyless prefix verification to check GitHub action, because this can be insecure. I'm referring to the shared workflows unwanted side effect that we discovered.

Talking about that, I think it would be even more relevant to have a section explaining why you should **not** use `KeylessPrefix`. It would be nice to add some example, like from our own policies that are actually signing policies in that way.

`GithubActions` or `KeylessPrefix` for verifying images. Read our
previous [blog post](https://www.kubewarden.io/blog/2022/05/verifying-image-signatures/) if you want to learn more about
how to verify container images with Sigstore using Kubewarden.

## Let's see it in action!

We want to verify the image `ghcr.io/raulcabello/app-example` which was built and signed inside a GitHub action
using this [GitHub Action](https://github.com/raulcabello/app-example/blob/master/.github/workflows/ci.yml).

Out of the box, GitHub Actions have a specially crafted environment that makes Sigstore keyless signing work in
a non-interactive way.
The signatures produced in this way contain unique secure information that allow us to identify the GitHub owner
(be it an individual or an organization) and the GitHub repository inside of which the GitHub Action has been
executed.

Starting from today, Kubewarden provides a convenient way to check signatures produced by GitHub actions.

For this example, a Kubernetes cluster with Kubewarden already installed is required. The installation process is
described in the [quick start guide](https://docs.kubewarden.io/quick-start).

> ### GitHubAction

Reviewer 2: ```suggestion
### GitHubActions
```

```yaml 
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicy
metadata:
  name: verify-image-signatures-policy
spec:
  module: registry://ghcr.io/kubewarden/policies/verify-image-signatures:v0.1.4
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
    - UPDATE
  mutating: true
  settings:
    signatures:
      - image: "ghcr.io/raulcabello/app-example:*" # match all tags 
        github_actions:
        - owner: "raulcabello"

>           repo: "app-example" #optional. if omited all apps from the owner will ve valid.

Reviewer 2: ```suggestion
          repo: "app-example" #optional. if omited all signatures performed in repos from the owner will be valid.
```

```

This policy verifies all containers with an image that is `ghcr.io/raulcabello/app-example`. It will accept

> containers that were signed inside a GitHub action for the owner `raulcabello` and the app `app-example`.

Reviewer 2: ```suggestion
containers that were signed inside a GitHub Action job, run under the owner `raulcabello` and the repo `app-example`.
```
Or something similar, so people understand we are validating things signed in job runs.

metadata:
  name: verify-image-signatures-policy
spec:
  module: registry://ghcr.io/kubewarden/policies/verify-image-signatures:v0.1.4
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
    - UPDATE
  mutating: true
  settings:
    signatures:
      - image: "ghcr.io/raulcabello/app-example:*" # match all tags 
        keyless_prefix:
          - issuer: "https://token.actions.githubusercontent.com"
            subject: "https://github.com/raulcabello/app-example/.github/workflows/ci.yml@refs/tags/" # match all tags
``` 

This will accept containers whose signature contains the issuer `https://token.actions.githubusercontent.com` and the 
subject starts with `https://github.com/raulcabello/app-example/.github/workflows/ci.yml@refs/tags/`. 

Like in the previous example this will accept the image `ghcr.io/raulcabello/app-example`, however we **don't** recommend using 
`KeylessPrefix` for GitHub actions validation. The reason is that `KeylessPrefix` will validate using just the subject, 
on the other hand `GithubActions` validates using the `github_workflow_repository` extension in the signature certificate. 

> We found out when signing our policies in a [shared workflow](https://github.com/kubewarden/github-actions/blob/v1/policy-release/action.yaml#L43) 

Reviewer 2: I would reword this and remove the "we found out" tbf. The thing is, the OIDC token is working as intended by design, but one should use the verification correctly. I would word is similar to:

"When GitHub creates the OIDC token used for the signatures, it sets `subject` as the URL containing the GHA workflow code, which doesn't necessarily match **where** it has run in the case of reusable workflows. This is by design. If one wants to check for the repo where the job was run, corresponding to that workflow code, GitHub has added a x502 certificate extension `github_workflow_repository` that contains it."

(also, notice is reusable workflow, not shared workflow).

I would also mention that in the future we intend to fail validation if KeylessPrefix has issuers that match `"https://token.actions.githubusercontent.com"`.

Author: done https://github.com/kubewarden/kubewarden.io/pull/116/commits/a2d9cd50f55120dd0526944f402ed4eb5213da5a
